### PR TITLE
408 circular route 210304

### LINKED ISSUE #408: Feature request: allow circular routes
I saw a review saying this is available on iPhone: allow planning of a circular route from start point, entering number of miles you wish to cycle.

### PR DESCRIPTION
**Pull request by HilaryN**
To fix #408.

I've only done Duration and Distance – not POI's yet.
It works from the map screen but not from the Plan a Route screen where you put in waypoints manually.

Terminology note:  circular and leisure route are the same thing and are used interchangeably, as are linear route and A-B route.

### Summary of changes

- Created activity for input of circular route details (distance or duration).
- Added extra API call for circular route. (Also added error parm to API calls.)
- Amended bits in between front and back end.
- Amended LiveRide replan route to work for circular route.

More detailed notes:
[Circular route notes 210305.pdf](https://github.com/cyclestreets/android/files/6090318/Circular.route.notes.210305.pdf)

Example json for circular/leisure route:
[json leisure.txt](https://github.com/cyclestreets/android/files/6090320/json.leisure.txt)

Test plan/log:
[Circular routes test plan 210305.pdf](https://github.com/cyclestreets/android/files/6090361/Circular.routes.test.plan.210305.pdf)

Scrappy picture of function calls:
Circular route fn calls

### CODE REVIEW

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/views/overlay/TapToRouteOverlay.kt
Comment (oliverlockwood) [inline libraries/cyclestreets-view/src/main/java/net/cyclestreets/views/overlay/TapToRouteOverlay.kt:285]: I would propose changing this string to: `Tap map to set Waypoint, or\nTap here for Circular Route`

Aside: we should probably move these texts into the languages XML so they can be internationalised

File: libraries/cyclestreets-view/src/main/res/values/strings.xml
Comment (oliverlockwood) [inline libraries/cyclestreets-view/src/main/res/values/strings.xml:36]: Any reason for making these new strings not translatable?  If it was just "I'm still developing" then fine, but we should change this when all other comments are addressed, and add placeholders (with the English values) in the different languages XML so they're ready to be translated.

File: libraries/cyclestreets-view/src/main/AndroidManifest.xml
Comment (oliverlockwood) [inline libraries/cyclestreets-view/src/main/AndroidManifest.xml:18]: If this _does_ need to be its own activity, why is it themed where others are not? and why is it within `views` package when other activities are not?

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/CycleStreetsRoutingTask.kt
Comment (oliverlockwood) [inline libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/CycleStreetsRoutingTask.kt:16]: I don't like that the interface change you've made forces us to put a magic number -1 here.
If it has to be so, please define a constant something like `NO_ITINTERARY = -1`, and use it both here and within the test in `doFetchRoute()`

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/RoutingTask.kt
Comment (oliverlockwood) [inline libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/RoutingTask.kt:39]: Don't throw raw `Exception`, please 🤮 
If you deliberately want to use a checked exception, then it must be a specific custom subclass, so we can explicitly handle it in a specific way.

As you're just using it to jump to the catch block, I'd accept throwing raw `RuntimeException`, if you don't want to define a custom exception.

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/RoutingTask.kt
Comment (oliverlockwood) [inline libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/RoutingTask.kt:46]: I absolutely cannot accept manual parsing of a JSON string, and neither should you 😉 

Suggest you instantiate an `ObjectMapper` at appropriate scope (the class?) and then either:
- define a POJO representing the JSON model, and parse into that, or:
- parse into ObjectNode, and then read the value of the `error` field, as explained in https://stackoverflow.com/a/26191723/2108057
`objectMapper.readValue(json, JourneyDomainObject::class.java)

**Review by oliverlockwood — CHANGES_REQUESTED** [review]
Some initial thoughts from a first pass.  I haven't looked at everything but I thought I'd rather give you some feedback now than wait until I've processed it all.

File: libraries/cyclestreets-view/src/main/res/values/strings.xml
Comment (HilaryN) [reply libraries/cyclestreets-view/src/main/res/values/strings.xml:36]: Being lazy, avoiding errors in the build for now.
I'll add them in.
btw Is there an easy way for a translator to see which strings still need translating, other than reading through them all?

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/CycleStreetsRoutingTask.kt
Comment (HilaryN) [reply libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/CycleStreetsRoutingTask.kt:16]: I'm pretty sure the -1 was already there in the previous code. (Probably in the FetchCycleStreetsRoutingTask.)  There were 2 fetchRoute functions and one just called the other, so it seemed tidier to combine them.

Happy to make a constant though.

File: cyclestreets.app/src/main/AndroidManifest.xml
Comment (HilaryN) [reply cyclestreets.app/src/main/AndroidManifest.xml:92]: Thanks for these comments.

What would you suggest otherwise, instead of the Activity?  

I originally thought to do an alert dialog, but the number of buttons (including cancel) exceeded the max.  (Bear in mind we still have to add in a button to select POI categories.)

File: libraries/cyclestreets-view/src/main/AndroidManifest.xml
Comment (HilaryN) [reply libraries/cyclestreets-view/src/main/AndroidManifest.xml:18]: I wanted to have the more modern theme of the back arrow at the top, which I don't think existed in earlier Android versions.
It wasn't clear to me how the classes are grouped in the packages, so I put it in the same place as the overlays as it's coming from the TapToRouteOverlay - I'm happy to move it to where the other activities are if you prefer.

File: libraries/cyclestreets-view/src/main/res/values/strings.xml
Comment (oliverlockwood) [reply libraries/cyclestreets-view/src/main/res/values/strings.xml:36]: Dunno, Google it and see 🙂

File: libraries/cyclestreets-view/src/main/AndroidManifest.xml
Comment (oliverlockwood) [reply libraries/cyclestreets-view/src/main/AndroidManifest.xml:18]: Perhaps you could include some screenshots of the new user journeys etc so I can see what it all looks like without running it locally, and then we can discuss.

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/CycleStreetsRoutingTask.kt
Comment (oliverlockwood) [reply libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/CycleStreetsRoutingTask.kt:16]: Indeed, we used an overload method for this. Outside the scope here makes it harder to see what's meant. And anyway, boy/girl scout rule of making things better as you go along, a descriptive constant would indeed be an improvement anyway 🙂

File: cyclestreets.app/src/main/AndroidManifest.xml
Comment (oliverlockwood) [reply cyclestreets.app/src/main/AndroidManifest.xml:92]: I'll ponder after seeing some screens. You may well be right, I just wanted to discuss

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/RoutingTask.kt
Comment (mvl22) [reply libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/RoutingTask.kt:46]: Yes, never parse JSON manually is a lesson we learnt :)

Comment (oliverlockwood) [reply libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/RoutingTask.kt:46]: My rule of thumb is always to ask myself ten times a day "am I doing something tedious which surely must be written better already in a library?"

File: libraries/cyclestreets-view/src/main/AndroidManifest.xml
Comment (HilaryN) [reply libraries/cyclestreets-view/src/main/AndroidManifest.xml:18]: Apologies, I should have included a link to the issues thread, where I've described the flow with screenshots:
https://github.com/cyclestreets/android/issues/408

Comment (HilaryN) [reply libraries/cyclestreets-view/src/main/AndroidManifest.xml:18]: I've added screenshots for the Distance tab and also for LiveRide / Replan route.

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/views/overlay/TapToRouteOverlay.kt
Comment (HilaryN) [reply libraries/cyclestreets-view/src/main/java/net/cyclestreets/views/overlay/TapToRouteOverlay.kt:285]: I guess I should do the same for "Tap map to set Waypoint\nTap here to Route" --> "Tap map to set Waypoint, or\nTap here to Route".

Yes, good point, I'll move the strings.  I'm trying to move them whenever I come across them, but I was probably too bogged down in other stuff at the time and then forgot! ( - and forgot to put in a //todo)

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/RoutingTask.kt
Comment (HilaryN) [reply libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/RoutingTask.kt:39]: I must admit, I only covered error handling and exceptions briefly in my reading so far, so this is the perfect cue for me to go and do some more in-depth reading.  Bear with me... :-)

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/views/overlay/TapToRouteOverlay.kt
Comment (HilaryN) [reply libraries/cyclestreets-view/src/main/java/net/cyclestreets/views/overlay/TapToRouteOverlay.kt]: I couldn't work out how to access the context easily from inside the enum class, so have done the context.getString outside the class.  From what I can see, the actionDesc for ALL_DONE doesn't get called, but I've done a check just in case.

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/RoutingTask.kt
Comment (HilaryN) [reply libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/RoutingTask.kt]: I've now defined a custom exception but not sure if this was what you had in mind?

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/RoutingTask.kt
Comment (HilaryN) [reply libraries/cyclestreets-view/src/main/java/net/cyclestreets/routing/RoutingTask.kt]: Using JsonNode seemed the simplest option for a single field.
Having "Error: " in the toast message seems superfluous, so I left that out of the toast message and just display the message itself.
Error message from server as toast

File: libraries/cyclestreets-view/src/main/res/values/strings.xml
Comment (HilaryN) [reply libraries/cyclestreets-view/src/main/res/values/strings.xml:36]: (Still todo: add in all the translation placeholders - will wait til end in case of any changed wording suggestions.)

**Comment by oliverlockwood** [comment]
@HilaryN I'll do a fuller review in a couple of days.  However one thing that occurs to me is you need to update `beta.txt` and `whatsnew.html` to describe the new features.  I've put a line (commented out) into `whatsnew.html` that you may choose to use - if you're happy with it, just comment it out and then add the same text without the HTML markup into the `beta.txt` file.
(I've updated your branch with the latest code from `master`, so you'll need to pull the changes down to your local working copy.)

**Comment by HilaryN** [comment]
> 
> 
> ... - if you're happy with it, just comment it out and then add the same text without the HTML markup into the `beta.txt` file.
> (I've updated your branch with the latest code from `master`, so you'll need to pull the changes down to your local working copy.)

Done this now, but before I push I just want to confirm that I should leave the text commented out in whatsnew?
Or was that a typo and should I _uncomment_ the text in whatsnew?

**Comment by oliverlockwood** [comment]
It was a typo, now corrected 🙂

File: libraries/cyclestreets-view/src/main/res/layout/activity_circular_route.xml
Comment (HilaryN) [inline libraries/cyclestreets-view/src/main/res/layout/activity_circular_route.xml:35]: In Landscape mode the tabs were squashed into the middle of the screen despite having "match_parent on the width". (Sorry I don't have a screenshot of that.)
tabMaxWidth="0dp" allows the tabs to expand to full screen width.

**Comment by HilaryN** [comment]
I've added the ViewModel as previously data entered by the user (Distance/Duration) was lost if the screen was rotated.

**Review by oliverlockwood — APPROVED** [review]
@HilaryN as you can see, I've added a number of commits to your changes, largely for maintainability.  The key things were:
- further improving error handling (although you had started using `ObjectMapper` for JSON parsing, you were still string-comparing a JSON string to see if it contained "error", which is bad practice)
- much simplification & clarification of the `CircularRouteActivity` / `CircularRouteViewModel` - enabled by using constants for the tab indexing approach, and separating the logic of "which tab am I on" from "what are the values for each tab", which in turn allowed us to drop the "storeValues" duplication
- tidying up of constants etc.

This is close to being ready to go.  In an ideal world, what I'd like to see is:
- the waypoints that are generated for a circular route being displayed on the map
- translations of the various strings we're adding.

However, as this goes into our beta pipeline, my view (if you're agreeable to it) is to merge this as-is and consider those aspects as improvements to be added subsequently.

On this basis I am approving the PR.  I shall hold off from merging until I hear from you about the above; also, I thought it would be polite to give you a chance to review the changes I've added on top of yours, to check you're happy with them before I merge them in under your name 😄

**Comment by HilaryN** [comment]
Thanks.  Yes, I'm happy with the changes.

Regarding displaying the waypoints, I decided against that as I thought it would unnecessarily clutter the screen and wouldn't add any value.  (See comments in #408 on 21 Feb.)

I guess for consistency they would need to be removed from the itinerary though.

File: libraries/cyclestreets-core/src/main/java/net/cyclestreets/api/client/RetrofitApiClient.java
Comment (mvl22) [reply libraries/cyclestreets-core/src/main/java/net/cyclestreets/api/client/RetrofitApiClient.java:153]: > the waypoints that are generated for a circular route being displayed on the map

Just to say that this isn't actually necessary for a circular leisure route. It's arguably perhaps an error that the API actually transmits waypoints other than start/finish waypoint.

Comment (HilaryN) [reply libraries/cyclestreets-core/src/main/java/net/cyclestreets/api/client/RetrofitApiClient.java:153]: But they are needed if a replan becomes necessary in LiveRide, if I remember correctly.

Comment (HilaryN) [reply libraries/cyclestreets-core/src/main/java/net/cyclestreets/api/client/RetrofitApiClient.java:153]: First 2 screenshots show LiveRide for a circular route.  3rd screenshot shows replan after going off route:
LiveRide replan

Comment (mvl22) [reply libraries/cyclestreets-core/src/main/java/net/cyclestreets/api/client/RetrofitApiClient.java:153]: Yes, that's true, so yes, I suppose we do need to transmit them.

However, the point remains that they don't need to be shown in the general route result, as they're really 'virtual' waypoints, since the user never set them.

Comment (oliverlockwood) [reply libraries/cyclestreets-core/src/main/java/net/cyclestreets/api/client/RetrofitApiClient.java:153]: OK, fine by me to leave out the waypoints then 👍 

btw I've just had a play with the circular route UI on my phone since the upgrade went out after I merged this, looking good.  Great contribution @HilaryN, thank you again.  I promise to respond promptly once you've had the chance to get the translations prepared!

**Pull request merged**
